src/euphorie/client/tests/test_survey.py

```diff
@@ -167,3 +167,107 @@ def testSetupContext_TwoSteps(self):
+            model.Risk(title=u'question 2', risk_id='2', zodb_path='a/c')
+        )
+        self.session.flush()
+
+        self.q1.postponed = False
+        self.q1.identification = "no"
+
+        self.survey.update_completion_percentage()
+        self.assertEqual(self.survey.completion_percentage, 50)
+
+    def test_two_thirds_complete(self):
+        self.mod1 = self.survey.addChild(
+            model.Module(title=u'module 1', module_id='1', zodb_path='a')
+        )
+        self.q1 = self.mod1.addChild(
+            model.Risk(title=u'question 1', risk_id='1', zodb_path='a/b')
+        )
+        self.q2 = self.mod1.addChild(
+            model.Risk(title=u'question 2', risk_id='2', zodb_path='a/c')
+        )
+        self.mod2 = self.survey.addChild(
+            model.Module(title=u'module 2', module_id='2', zodb_path='k')
+        )
+        self.q21 = self.mod1.addChild(
+            model.Risk(title=u'question 3', risk_id='3', zodb_path='k/l')
+        )
+        self.session.flush()
+
+        self.q1.postponed = False
+        self.q1.identification = "n/a"
+        self.q21.postponed = False
+        self.q21.identification = "yes"
+
+        self.survey.update_completion_percentage()
+        self.assertEqual(self.survey.completion_percentage, 67)
+
+    def test_optional_module(self):
+        self.mod1 = self.survey.addChild(
+            model.Module(title=u'module 1', module_id='1', zodb_path='a')
+        )
+        self.q1 = self.mod1.addChild(
+            model.Risk(title=u'question 1', risk_id='1', zodb_path='a/b')
+        )
+        self.mod2 = self.survey.addChild(
+            model.Module(title=u'module 2', module_id='2', zodb_path='k')
+        )
+        self.q2 = self.mod2.addChild(
+            model.Risk(title=u'question 2', risk_id='2', zodb_path='k/c')
+        )
+        self.q1.postponed = False
+        self.q1.identification = "yes"
+        self.mod2.skip_children = True
+
+        self.survey.update_completion_percentage()
+        self.assertEqual(self.survey.completion_percentage, 100)
+
+    def test_optional_submodule(self):
+        self.mod1 = self.survey.addChild(
+            model.Module(title=u'module 1', module_id='1', zodb_path='a')
+        )
+        self.q1 = self.mod1.addChild(
+            model.Risk(title=u'question 1', risk_id='1', zodb_path='a/b')
+        )
+        self.mod2 = self.survey.addChild(
+            model.Module(title=u'module 2', module_id='2', zodb_path='k')
+        )
+        self.mod21 = self.mod2.addChild(
+            model.Module(title=u'module 3', module_id='3', zodb_path='k/c')
+        )
+        self.q2 = self.mod21.addChild(
+            model.Risk(title=u'question 2', risk_id='2', zodb_path='k/c/u')
+        )
+        self.q1.postponed = False
+        self.q1.identification = "yes"
+        self.mod21.skip_children = True
+
+        self.survey.update_completion_percentage()
+        self.assertEqual(self.survey.completion_percentage, 100)
```

Comment:
Please add a test that proves that if the module mod2 is optional and false, that the risks in submodule m21 are not counted.